### DISCUSSION TITLE
Consider renaming `Deterministic` to `store_in_trace`

### DISCUSSION THREAD

**Discussion by ricardoV94** [pymc-devs/pymc3#7209] [Ideas]
### Description

`Deterministic` does not convey the meaning of what it does. It stores an arbitrary computation (including stochastic nodes) in the trace, between sampled points.

From experience with Discourse users ([most recent example](https://discourse.pymc.io/t/why-would-scipy-numpy-wrapped-in-as-op-cause-much-faster-sampling-than-using-pytensor-operations/11975/3?u=ricardov94)), I think `Deterministic` is a source of confusion. Some users feel they need it for everything that is not a `pm.Distribution`. This is 
 1. Not true,
 2. A potential waste of computation and [slowdown](https://discourse.pymc.io/t/parallelizing-chains-with-custom-likelihood-on-multiple-cores/11504/18) 
 3. Can make the model API seem quite "uglier/verbose" than it actually is. 

We should keep the `Deterministic` as an alias for quite some time, but start using the new name in docs and examples.

Alternative names welcome

**Comment by cuchoi** [comment]
That makes a lot of sense to me... are there any other uses for pm.Deterministic? We could put a deprecation warning on pm.Deterministic, but that might be a bit noisy given its adoption level.

**Comment by ricardoV94** [comment]
Storing in trace and showing in graphviz are the only uses I can think of.

**Comment by fonnesbeck** [comment]
Perhaps something like `store_values`? If you are using VI, there will not be a trace *per se*, so may as well be agnostic there.

**Comment by ricardoV94** [comment]
Short would also be nice, maybe something like `record`?

**Comment by fonnesbeck** [comment]
Was also thinking about capitalization/camel case here. I assume we will also need to pass a name, so might be odd to have these using a lower case/underscore convention while everything else in the model is camel cased. The camel case convention sort of implies that it is a first-class variable in the model, which I like.

`theta = store_values('theta', alpha + beta * X)` or `theta = record('theta', alpha + beta * X)`

versus

`theta = StoreValues('theta', alpha + beta * X)` or `theta = Record('theta', alpha + beta * X)`

**Comment by fonnesbeck** [comment]
Alternative names:

- `Memoize`
- `Track`
- `Register`

feel free to add others if anyone has ideas.

**Comment by ricardoV94** [comment]
Not `Register`, because we "register" all variables. Others sound good.

Does github have polls?

**Comment by fonnesbeck** [comment]
Here's a quick poll. Use the associated emoji to vote.

- `Memoize` 🎉
- `Track` ❤️
- `Store` 🚀
- `Record` 😄
- none of the above 👎

**Comment by twiecki** [comment]
<img width="735" alt="image" src="https://github.com/pymc-devs/pymc/assets/674200/cb68ceaf-3031-4fc3-aa15-ffef9ccfbd33">

**Comment by fonnesbeck** [comment]
Its a virtual tie internally, so let's go with `Record`.

**Comment by twiecki** [comment]
Final results
<img width="757" alt="image" src="https://github.com/pymc-devs/pymc/assets/674200/443ead53-5893-4eaf-a786-60d0186ad986">

**Comment by ricardoV94** [comment]
This user was getting OOM problems because of the behavior of Deterministic: https://github.com/pymc-devs/pymc/issues/6852

OTOH calling something Deterministic sounds nice for targeting by the `do` and `observe` functionality that will soon move from pymc-experimental to pymc. That makes me a bit reluctant about the renaming

**Comment by twiecki** [comment]
I don't follow.

**Comment by ricardoV94** [comment]
```python
with pm.Model() as m:
  ...
  x = pm.Deterministic("x”, ...)
  ...
```
```python
new_m = observe(m, {x: data})
```
```python
new_m = do(m, {x: intervention})
```

Is more intuitive when `x` was defined as a "Deterministic" vs as a "store_in_trace".

**Comment by ricardoV94** [comment]
https://discourse.pymc.io/t/out-of-memory-during-transforming-variables-step-using-numpyro-increasing-postprocessing-chunks-doesnt-seems-to-help-what-should-i-do/13061/5?u=ricardov94

**Comment by twiecki** [comment]
> ```python
> with pm.Model() as m:
>   ...
>   x = pm.Deterministic("x”, ...)
>   ...
> ```
> 
> ```python
> new_m = observe(m, {x: data})
> ```
> 
> ```python
> new_m = do(m, {x: intervention})
> ```
> 
> Is more intuitive when `x` was defined as a "Deterministic" vs as a "store_in_trace".

I don't feel the same holds quite true for `Record()`. Another way would be `NamedDeterministic()/NamedDet()`, or `StoredDeterministic()/StoredDet()`.

**Comment by trendelkampschroer** [comment]
Thanks for the great package and making it freely available. It's a pleasure learning Beyesian data analysis through the many great examples and `pymc`s modeling language.

In addition emphasising the potential caveats of using `pm.Deterministic` in the examples, e.g [here](https://www.pymc.io/projects/docs/en/stable/learn/core_notebooks/pymc_overview.html#id1) would be helpful to new users.

Especially for hierarchical models, cf. [here](https://www.pymc.io/projects/examples/en/latest/generalized_linear_models/GLM-simpsons-paradox.html) lead me to include "deterministics" into models and end up with unnecessarily large traces and compute times for e.g. `arviz.summary(trace)`.

Also consider the usage of 'deterministics' [here](https://www.pymc.io/projects/docs/en/stable/learn/core_notebooks/pymc_overview.html#id1) which might also be confusing - at least when starting with `pymc`.
